lib/ruby_lsp/requests/request.rb

```diff
@@ -0,0 +1,36 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # :nodoc:
+    class Request
+      extend T::Sig
+      extend T::Generic
+
+      abstract!
+
+      sig { abstract.returns(T.anything) }
+      def response; end
+    end
+
+    # :nodoc:
+    class ListenerBasedRequest < Request
+      extend T::Sig
+      extend T::Generic
+
+      ResponseType = type_member
+
+      abstract!
+
+      sig { params(listeners: T::Array[Listener[ResponseType]]).void }
+      def initialize(listeners)
+        super()
+        @listeners = listeners
+      end
```

Comment:
Instead of initializing a request with a list of listeners, could we instantiate the listeners inside the request itself? Is there any value I'm missing with making the listeners an initialization parameter?

This is what I'm thinking:

```ruby
module Requests
  class CodeLens < Request
    def initialize(...)
      @listeners = instantiate_all_listeners
    end

    def run
      dispatcher = Prism::Dispatcher.new
      # ...
      response
    end
  end
end
```

In this case, we wouldn't even need the `ListenerBasedRequest` class.